In [14]:
import pandas as pd

In [15]:
import io
df2 = pd.read_csv(r'C:\Users\pedro\Desktop\Data Science\python_intro\Base_de_Datos_Proyecto1\sales_predictions.csv')

In [16]:
df2.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

1) ¿Cuántas filas y columnas?

In [17]:
df2.shape

(8523, 12)

Filas, Colummnas\
8523, 12

2) ¿Cuáles son los tipos de datos en cada variable?

In [18]:
df2.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

3) ¿Hay duplicados? Si es el caso, eliminen algunos duplicados.

In [19]:
df2.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
8518    False
8519    False
8520    False
8521    False
8522    False
Length: 8523, dtype: bool

In [20]:
df2.duplicated().sum()

0

No hay duplicados.

4) Identifiquen los valores faltantes.

In [21]:
df2.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

5)Decide cómo abordar los valores faltantes y cómo hacerlo. (Esto requiere sus criterios, así que expliquen su elección).

In [22]:
df2

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0              FDA15        9.300          Low Fat         0.016047   
1              DRC01        5.920          Regular         0.019278   
2              FDN15       17.500          Low Fat         0.016760   
3              FDX07       19.200          Regular         0.000000   
4              NCD19        8.930          Low Fat         0.000000   
...              ...          ...              ...              ...   
8518           FDF22        6.865          Low Fat         0.056783   
8519           FDS36        8.380          Regular         0.046982   
8520           NCJ29       10.600          Low Fat         0.035186   
8521           FDN46        7.210          Regular         0.145221   
8522           DRG01       14.800          Low Fat         0.044878   

                  Item_Type  Item_MRP Outlet_Identifier  \
0                     Dairy  249.8092            OUT049   
1               Soft Drinks   48.2692            OUT018   
2                      Meat  141.6180            OUT049   
3     Fruits and Vegetables  182.0950            OUT010   
4                 Household   53.8614            OUT013   
...                     ...       ...               ...   
8518            Snack Foods  214.5218            OUT013   
8519           Baking Goods  108.1570            OUT045   
8520     Health and Hygiene   85.1224            OUT035   
8521            Snack Foods  103.1332            OUT018   
8522            Soft Drinks   75.4670            OUT046   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                          1999      Medium               Tier 1   
1                          2009      Medium               Tier 3   
2                          1999      Medium               Tier 1   
3                          1998         NaN               Tier 3   
4                          1987        High               Tier 3   
...                         ...         ...                  ...   
8518                       1987        High               Tier 3   
8519                       2002         NaN               Tier 2   
8520                       2004       Small               Tier 2   
8521                       2009      Medium               Tier 3   
8522                       1997       Small               Tier 1   

            Outlet_Type  Item_Outlet_Sales  
0     Supermarket Type1          3735.1380  
1     Supermarket Type2           443.4228  
2     Supermarket Type1          2097.2700  
3         Grocery Store           732.3800  
4     Supermarket Type1           994.7052  
...                 ...                ...  
8518  Supermarket Type1          2778.3834  
8519  Supermarket Type1           549.2850  
8520  Supermarket Type1          1193.1136  
8521  Supermarket Type2          1845.5976  
8522  Supermarket Type1           765.6700  

[8523 rows x 12 columns]

# **Outlet_Size**

In [26]:
df2.groupby('Outlet_Identifier')['Outlet_Size'].count()

Outlet_Identifier
OUT010      0
OUT013    932
OUT017      0
OUT018    928
OUT019    528
OUT027    935
OUT035    930
OUT045      0
OUT046    930
OUT049    930
Name: Outlet_Size, dtype: int64

Los Outlet: OUT010,OUT017,OUT045. Carecen de tamano

In [42]:
df2.groupby(['Outlet_Identifier','Outlet_Type'])['Outlet_Size'].max()

Outlet_Identifier  Outlet_Type      
OUT010             Grocery Store           NaN
OUT013             Supermarket Type1      High
OUT017             Supermarket Type1       NaN
OUT018             Supermarket Type2    Medium
OUT019             Grocery Store         Small
OUT027             Supermarket Type3    Medium
OUT035             Supermarket Type1     Small
OUT045             Supermarket Type1       NaN
OUT046             Supermarket Type1     Small
OUT049             Supermarket Type1    Medium
Name: Outlet_Size, dtype: object

In [43]:
Frequency_Size= df2.groupby('Outlet_Type')['Outlet_Size'].value_counts()
Frequency_Size

Outlet_Type        Outlet_Size
Grocery Store      Small           528
Supermarket Type1  Small          1860
                   High            932
                   Medium          930
Supermarket Type2  Medium          928
Supermarket Type3  Medium          935
Name: Outlet_Size, dtype: int64

El unico valor que  se repite en la categoria Grocery Store es 'Small'
El valor que mas se repite en la categoria Supermarket Type 1 'Small'
Basandonos en este analisis podemos imputar el Size Small para los tipos de Outlet 'Supermarket Type1' y 'Grocery Store '


In [50]:
df2['Outlet_Size'].fillna('Small', inplace=True)

In [51]:
df2.groupby(['Outlet_Identifier','Outlet_Type'])['Outlet_Size'].max()

Outlet_Identifier  Outlet_Type      
OUT010             Grocery Store         Small
OUT013             Supermarket Type1      High
OUT017             Supermarket Type1     Small
OUT018             Supermarket Type2    Medium
OUT019             Grocery Store         Small
OUT027             Supermarket Type3    Medium
OUT035             Supermarket Type1     Small
OUT045             Supermarket Type1     Small
OUT046             Supermarket Type1     Small
OUT049             Supermarket Type1    Medium
Name: Outlet_Size, dtype: object

# **Item_Weight**

In [52]:
df2.sort_values('Item_Weight')

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
4430           FDP40        4.555          Regular         0.034357   
7808           FDP40        4.555          Regular         0.034329   
3489           FDP40        4.555          Regular         0.034351   
4400           FDP40        4.555          Regular         0.034411   
7984           DRE12        4.590          Low Fat         0.000000   
...              ...          ...              ...              ...   
43             FDC02       21.350          Low Fat         0.069103   
2802           FDC02       21.350          Low Fat         0.068765   
483            FDC02       21.350          Low Fat         0.115195   
2368           FDC02       21.350          Low Fat         0.068809   
4468           FDC02       21.350          Low Fat         0.068822   

         Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
4430  Frozen Foods  112.6544            OUT046                       1997   
7808  Frozen Foods  110.1544            OUT013                       1987   
3489  Frozen Foods  112.7544            OUT035                       2004   
4400  Frozen Foods  111.3544            OUT049                       1999   
7984   Soft Drinks  111.1860            OUT045                       2002   
...            ...       ...               ...                        ...   
43          Canned  259.9278            OUT018                       2009   
2802        Canned  260.4278            OUT013                       1987   
483         Canned  258.3278            OUT010                       1998   
2368        Canned  258.5278            OUT035                       2004   
4468        Canned  258.3278            OUT046                       1997   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
4430       Small               Tier 1  Supermarket Type1          2684.5056  
7808        High               Tier 3  Supermarket Type1          1230.3984  
3489       Small               Tier 2  Supermarket Type1          1789.6704  
4400      Medium               Tier 1  Supermarket Type1          1342.2528  
7984       Small               Tier 2  Supermarket Type1          1245.0460  
...          ...                  ...                ...                ...  
43        Medium               Tier 3  Supermarket Type2          6768.5228  
2802        High               Tier 3  Supermarket Type1          3644.5892  
483        Small               Tier 3      Grocery Store           520.6556  
2368       Small               Tier 2  Supermarket Type1          5206.5560  
4468       Small               Tier 1  Supermarket Type1          7028.8506  

[8523 rows x 12 columns]

No existe una relación entre si falta un punto de datos y cualquier valor en el conjunto de datos (faltante u observado). Los datos faltantes son solo un subconjunto aleatorio de los datos. La ausencia no tiene nada que ver con ninguna otra variable. Por lo que para generar los valores faltantes se utilizara el proceso de imputacion de la media por 'Item_Type'

In [53]:
df2.sort_values('Item_Type')

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
4261           FDY36    12.300000          Low Fat         0.009425   
5387           FDS60    20.850000          Low Fat         0.032581   
5363           FDY48    14.000000          Low Fat         0.023783   
5362           FDZ48    17.750000          Low Fat         0.076114   
5354           FDN24    12.277108          Low Fat         0.198317   
...              ...          ...              ...              ...   
7353           FDC59    16.700000              reg         0.091438   
7363           FDE23    17.600000          Regular         0.053482   
136            FDH35    13.690731          Low Fat         0.059957   
6723           FDG47    12.800000          Low Fat         0.069606   
5237           FDB11    16.000000          Low Fat         0.060971   

          Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
4261   Baking Goods   74.8380            OUT049                       1999   
5387   Baking Goods  180.0660            OUT018                       2009   
5363   Baking Goods  104.0332            OUT045                       2002   
5362   Baking Goods  111.4544            OUT045                       2002   
5354   Baking Goods   54.3956            OUT019                       1985   
...             ...       ...               ...                        ...   
7353  Starchy Foods   64.3168            OUT010                       1998   
7363  Starchy Foods   47.8060            OUT017                       2007   
136   Starchy Foods  165.4526            OUT027                       1985   
6723  Starchy Foods  261.9252            OUT035                       2004   
5237  Starchy Foods  226.2404            OUT045                       2002   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
4261      Medium               Tier 1  Supermarket Type1          1537.9980  
5387      Medium               Tier 3  Supermarket Type2          2157.1920  
5363       Small               Tier 2  Supermarket Type1          1435.4648  
5362       Small               Tier 2  Supermarket Type1          1006.6896  
5354       Small               Tier 1      Grocery Store           272.9780  
...          ...                  ...                ...                ...  
7353       Small               Tier 3      Grocery Store           127.8336  
7363       Small               Tier 2  Supermarket Type1          1071.9380  
136       Medium               Tier 3  Supermarket Type3          5262.4832  
6723       Small               Tier 2  Supermarket Type1          4984.1788  
5237       Small               Tier 2  Supermarket Type1          4725.8484  

[8523 rows x 12 columns]

In [54]:
df2['Item_Weight'].fillna(df2.groupby('Item_Type')['Item_Weight'].transform('mean'), inplace = True)

In [55]:
df2.sort_values('Item_Type')

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
4261           FDY36    12.300000          Low Fat         0.009425   
5387           FDS60    20.850000          Low Fat         0.032581   
5363           FDY48    14.000000          Low Fat         0.023783   
5362           FDZ48    17.750000          Low Fat         0.076114   
5354           FDN24    12.277108          Low Fat         0.198317   
...              ...          ...              ...              ...   
7353           FDC59    16.700000              reg         0.091438   
7363           FDE23    17.600000          Regular         0.053482   
136            FDH35    13.690731          Low Fat         0.059957   
6723           FDG47    12.800000          Low Fat         0.069606   
5237           FDB11    16.000000          Low Fat         0.060971   

          Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
4261   Baking Goods   74.8380            OUT049                       1999   
5387   Baking Goods  180.0660            OUT018                       2009   
5363   Baking Goods  104.0332            OUT045                       2002   
5362   Baking Goods  111.4544            OUT045                       2002   
5354   Baking Goods   54.3956            OUT019                       1985   
...             ...       ...               ...                        ...   
7353  Starchy Foods   64.3168            OUT010                       1998   
7363  Starchy Foods   47.8060            OUT017                       2007   
136   Starchy Foods  165.4526            OUT027                       1985   
6723  Starchy Foods  261.9252            OUT035                       2004   
5237  Starchy Foods  226.2404            OUT045                       2002   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
4261      Medium               Tier 1  Supermarket Type1          1537.9980  
5387      Medium               Tier 3  Supermarket Type2          2157.1920  
5363       Small               Tier 2  Supermarket Type1          1435.4648  
5362       Small               Tier 2  Supermarket Type1          1006.6896  
5354       Small               Tier 1      Grocery Store           272.9780  
...          ...                  ...                ...                ...  
7353       Small               Tier 3      Grocery Store           127.8336  
7363       Small               Tier 2  Supermarket Type1          1071.9380  
136       Medium               Tier 3  Supermarket Type3          5262.4832  
6723       Small               Tier 2  Supermarket Type1          4984.1788  
5237       Small               Tier 2  Supermarket Type1          4725.8484  

[8523 rows x 12 columns]

6) Confirmen que no hay valores faltantes después de abordarlos.



In [56]:
df2.isna().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

Encuentren y arreglen alguna categoría inconsistente de datos

In [67]:
def unique_values (dataframe):
    unive_values_r=''
    for columns in range(dataframe):
      unive_values_r= dataframe[columns].unique
    return unive_values_r

In [69]:
df2['Item_Identifier'].unique()

array(['FDA15', 'DRC01', 'FDN15', ..., 'NCF55', 'NCW30', 'NCW05'],
      dtype=object)

In [70]:
df2['Item_Weight'].unique()

array([ 9.3       ,  5.92      , 17.5       , 19.2       ,  8.93      ,
       10.395     , 13.65      , 12.98787955, 16.2       , 11.8       ,
       18.5       , 15.1       , 17.6       , 16.35      ,  9.        ,
       11.40032787, 13.35      , 18.85      , 12.27710821, 14.6       ,
       13.85      , 13.        ,  7.645     , 11.65      ,  5.925     ,
       12.30570501, 19.25      , 18.6       , 18.7       , 17.85      ,
       10.        , 13.22476938,  8.85      ,  9.8       , 13.6       ,
       21.35      , 12.15      ,  6.42      , 19.6       , 15.85      ,
       13.4260689 ,  7.39      , 10.195     ,  9.895     , 10.895     ,
        7.905     ,  9.195     ,  8.365     ,  7.97      , 12.86706128,
       17.7       , 19.35      ,  8.645     , 15.6       , 18.25      ,
        7.855     ,  7.825     ,  8.39      , 12.85      , 19.        ,
        5.905     ,  7.76      , 16.75      , 13.3847365 , 15.5       ,
        6.055     , 12.76820225,  6.305     , 20.85      , 20.75

In [85]:
df2['Item_Fat_Content'].unique()

array(['Low Fat', 'Regular', 'low fat', 'LF', 'reg', nan], dtype=object)

In [108]:
df2.Item_Fat_Content= df2.Item_Fat_Content.replace({"low fat": "Low Fat", "LF": "Low Fat", "reg": "Regular"})


In [110]:
df2['Item_Fat_Content'].unique()

array(['Low Fat', 'Regular', nan], dtype=object)

In [111]:
df2['Item_Type'].unique()

array(['Dairy', 'Soft Drinks', 'Meat', 'Fruits and Vegetables',
       'Household', 'Baking Goods', 'Snack Foods', 'Frozen Foods',
       'Breakfast', 'Health and Hygiene', 'Hard Drinks', 'Canned',
       'Breads', 'Starchy Foods', 'Others', 'Seafood', nan], dtype=object)

In [112]:
df2['Outlet_Establishment_Year'].unique()

array([1999., 2009., 1998., 1987., 1985., 2002., 2007., 1997., 2004.,
         nan])

In [113]:
df2['Outlet_Identifier'].unique()

array(['OUT049', 'OUT018', 'OUT010', 'OUT013', 'OUT027', 'OUT045',
       'OUT017', 'OUT046', 'OUT035', 'OUT019', nan], dtype=object)

Solo se identifico inconsistencia en la columna 'Item_Fat_Content'

8) Para cualquier columna numérica, obtengan las estadísticas resumidas para cada uno (mínimo, máximo y media)

In [115]:
df2.describe()

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  8523.000000      8523.000000  8523.000000                8523.000000   
mean     12.857890         0.066132   140.992782                1997.831867   
std       4.232804         0.051598    62.275067                   8.371760   
min       4.555000         0.000000    31.290000                1985.000000   
25%       9.310000         0.026989    93.826500                1987.000000   
50%      12.867061         0.053931   143.012800                1999.000000   
75%      16.000000         0.094585   185.643700                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8523.000000  
mean         2181.288914  
std          1706.499616  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3101.296400  
max         13086.964800